code/Control/UserDefinedFormController.php

```diff
@@ -282,6 +292,12 @@ public function process($data, $form)
                         $file->UserFormUpload = UserFormFileExtension::USER_FORM_UPLOAD_TRUE;
                         $file->write();
 
+                        if ($currentStage) {
+                            Versioned::set_stage($currentStage);
+                        } else {
+                            Versioned::reset();
```

Comment:
I don't think this is the correct place for this.  If we're concerned about the null return in Versioned::get_stage then it would make sense to update it in Versioned::withReadingMode rather here which would only update it in a single place.

 Instead could you please just update this to Versioned::withReadingMode as mentioned here - https://github.com/silverstripe/silverstripe-userforms/pull/1144#discussion_r874371165 - we use this method in multiple other places in the codebase so it's safe to use here

Reply:
I deliberately moved all changes for Stage to this level, since I need to set the Draft version only for the files in the folder, not for the folder itself.
Therefore, I need to call the set_stage method only after the file has been written.

Reply:
The way this code is currently written, if it needs to create a new folder, it will create the folder in draft only mode since `Versioned::set_stage($this->config()->get('file_upload_stage'));` (draft) is called directly before `UserDefinedFormController::setProtectedFolder()` which calls `UserDefinedFormAdmin::getFormSubmissionFolder()` which will create a folder if required.  

So `Versioned::set_stage($this->config()->get('file_upload_stage'));` (or better `Versioned::withReadingMode()`) also needs to be moved after checking/creating the folder

Reply:
Probably I misunderstood something. I'm just trying to follow this requirement:
- Files will default to be uploaded in draft state all the time
I'll move Versioned::set_stage($this->config()->get('file_upload_stage')); after  setProtectedFolder().

Reply:
Probably I misunderstood something. I'm just trying to follow this requirement:
- Files will default to be uploaded in draft state all the time
I'll move Versioned::set_stage($this->config()->get('file_upload_stage')); after  setProtectedFolder().

Reply:
Something along these lines

```php
    if (!$field->getFolderExists()) {
        $field->createProtectedFolder();
    }

    $doReturn = false;
    Versioned::withReadingMode(function() use ($form, $field, $doReturn) {
        $stage = Injector::inst()->get(self::class)->config()->get('file_upload_stage');
        Versioned::set_stage($this, $stage);
        $folderName = $field->getFormField()->getFolderName();
        // create the file from post data
        $upload = Upload::create();
        try {
            $upload->loadIntoFile($_FILES[$field->Name], null, $folderName);
        } catch (ValidationException $e) {
            $validationResult = $e->getResult();
            foreach ($validationResult->getMessages() as $message) {
                $form->sessionMessage($message['message'], ValidationResult::TYPE_ERROR);
            }
            Controller::curr()->redirectBack();
            $doReturn = true;
            return;
        }
        /** @var AssetContainer|File $file */
        $file = $upload->getFile();
        $file->ShowInSearch = 0;
        $file->UserFormUpload = UserFormFileExtension::USER_FORM_UPLOAD_TRUE;
        $file->write();
    });
    if ($doReturn) {
        return;
    }
```